rfcs/text/0013_saved_object_migrations.md

```diff
@@ -0,0 +1,527 @@
+> perform a migration at the same time. E.g. if a v7.8.0 and a v7.9.0 node
+> attempts to simultaneously migrate a 7.6.0 index the result will be an
+> inconsistent state. This is an unlikely scenario in practise and therefore
+> an acceptable limitation. In 8.x we can ensure that only the latest node
+> completes the migration. 
+
+## 4.5 Changes in 8.0
+1. Migrations will be in-place and re-use the same index for minor and patch
+   upgrades. 
+2. Only allow breaking field mapping changes (which requires creating a new
+   index) in a major.
+3. All update operations should use optimistic concurrency control.
+
+### 4.5.1 Migration algorithm (8.0):
+
+1. On startup, find the index to which the `.kibana` alias points,
+   `.kibana_n`, and check if any documents have an outdated `migrationVersion`
+   indicating that a migration needs to be performed.
+2. If there are documents with newer `migrationVersion` numbers, exit Kibana
+   with a fatal error.
+3. If there are outdated documents, migrate these in batches:
+   1. Read a batch of outdated documents from `.kibana_n`.
+   2. Transform documents by applying the migration functions.
+   3. Update the document batch in the same index using optimistic concurrency
+      control. 
+   4. If a batch fails due to an update version mismatch, repeat the entire
+      process.
+4. Once all documents are up to date, the migration is complete and Kibana can
+   start serving traffic.
+
+Advantages:
+- Not duplicating all documents into a new index will speed up migrations and
+  reduce the downtime window. This will be especially important for the future
+  requirement to support > 10k or > 100k documents.
+- We can check the health of an existing index before starting the migration,
+  but we cannot detect what kind of failures might occur while creating a new
+  index. Whereas retrying migrations will eventually recover from the errors
+  in (3.3), re-using an index allows us to detect these problems before trying
+  and avoid errors like (3.3.1) altogether.
+- Single index to backup instead of “index pattern” that matches any
+  `.kibana_n`.
+- Simplifies Kibana system index Elasticsearch plugin since it needs to work
+  on one index per "tenant".
+- Can prevent data loss when mixed versions of Kibana are both attempting a
+  migration using optimistic concurrency control.
+
+Drawbacks:
+- Cannot make breaking mapping changes (even though it was possible, we have not
+  introduced a breaking mapping change during 7.x).
+- Rollback is only possible by restoring a snapshot which requires educating
+  users to ensure that they don't rely on `.kibana_n` indices as backups.
+  (Apart from the need to educate users, snapshot restores provide many
+  benefits).
+- It narrows the second restriction under (4.3.1) even further: migrations
+  cannot rely on any state that could change as part of a migration because we
+  can no longer use the previous index as a snapshot of unmigrated state.
```

Comment:
Does this mean migrations won't have access to read data from other saved objects? I can't remember the specifics now... but I thought we needed that functionality for some reason.  @timroes you remember?

Reply:
We can still read other saved objects but they might or might not be migrated already. Whereas we previously had a snapshot that we knew would remain unmigrated forever. So I don't think this will affect https://github.com/elastic/kibana/issues/34996

Reply:
Though I have my doubts whether we can support https://github.com/elastic/kibana/issues/34996 generically. We can probably expose this functionality, but every migration that is written with such a pre-migration state collector will have to be very carefully thought through since the interactions with other migrations that potentially re-write id's make it very hard to reason about the behaviour during the migration and during saved object imports.

So we will probably have to consider these kinds of migrations on a case-by-case basis to really say if it's feasible.

Reply:
Will the API types take this into account so it's obvious that the type might be the latest shape or the shape from the last version? E.g. the saved object getter would return something like `MyObject7_1` | `MyObject7_2`?

I'm worried that this could cause hard to debug issues if not clearly indicated by the API that the type is not guaranteed to be the latest.  Consider:

1. Time T1: Developer of Bar plugin migrates Bar to Bar1, adding property `zed` to Bar1 type.

2. Time T1: Developer of Foo plugin, migrates Foo to Foo1, looks up other saved object, Bar1, to do so, and accesses property `zed`.

Will the API warn Foo developer that Bar1 type might actually _not_ have property `zed` on it?

Reply:
The reason why we discussed this, and that feature would be nice is the embedding in dashboard functionality. Currently we have no way of migrating an dashboard using references for panels to a one where the visualization is stored embedded in the dashboard, because that would require us to load that visualization saved object while migrating the dashboard. The other aspect was removing of the markdown visualization and input controls. We're planning to make them more like "widgets" on the dashboard and not part of visualizations. But to migrate that we'd also need to load an existing visualization object during the dashboard migration to figure out the content of the markdown visualization to convert it to a widget.

I'd say neither of those is at the moment really critical, or rather.. we didn't expect to be able to do either of these. While nice to haves, I wouldn't let that use-case block something else right now.